File: ontrack-model/src/main/java/net/nemerosa/ontrack/model/settings/LDAPSettings.java
package net.nemerosa.ontrack.model.settings;

import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.experimental.Wither;
import net.nemerosa.ontrack.model.form.Form;
import net.nemerosa.ontrack.model.form.Password;
import net.nemerosa.ontrack.model.form.Text;
import net.nemerosa.ontrack.model.form.YesNo;

@Data
@AllArgsConstructor(access = AccessLevel.PUBLIC)
public class LDAPSettings {

    private final boolean enabled;
    private final String url;
    private final String searchBase;
    private final String searchFilter;
    private final String user;
    private final String password;
    @Wither
    private final String fullNameAttribute;
    @Wither
    private final String emailAttribute;
    @Wither
    private final String groupAttribute;
    @Wither
    private final String groupFilter;

    public Form form() {
        return Form.create()
                .with(
                        YesNo.of("enabled")
                                .label("Enable LDAP authentication")
                                .value(enabled)
                )
                .with(
                        Text.of("url")
                                .visibleIf("enabled")
                                .label("URL")
                                .help("URL to the LDAP server")
                                .value(url)
                )
                .with(
                        Text.of("user")
                                .visibleIf("enabled")
                                .label("User")
                                .help("Name of the user used to connect to the LDAP server")
                                .optional()
                                .value(user)
                )
                .with(
                        Password.of("password")
                                .visibleIf("enabled")
                                .label("Password")
                                .help("Password of the user used to connect to the LDAP server")
                                .optional()
                                .value("") // Password never sent to the client
                )
                .with(
                        Text.of("searchBase")
                                .visibleIf("enabled")
                                .label("Search base")
                                .help("Query to get the user")
                                .optional()
                                .value(searchBase)
                )
                .with(
                        Text.of("searchFilter")
                                .visibleIf("enabled")
                                .label("Search filter")
                                .help("Filter on the user query")
                                .optional()
                                .value(searchFilter)
                )
                .with(
                        Text.of("fullNameAttribute")
                                .visibleIf("enabled")
                                .label("Full name attribute")
                                .help("Name of the attribute that contains the full name of the user")
                                .optional()
                                .value(fullNameAttribute)
                )
                .with(
                        Text.of("emailAttribute")
                                .visibleIf("enabled")
                                .label("Email attribute")
                                .help("Name of the attribute that contains the email of the user")
                                .optional()
                                .value(emailAttribute)
                )
                .with(
                        Text.of("groupAttribute")
                                .visibleIf("enabled")
                                .label("Group attribute")
                                .help("Name of the attribute that contains the groups the user belongs to (defaults to memberOf)")
                                .optional()
                                .value(groupAttribute)
                )
                .with(
                        Text.of("groupFilter")
                                .visibleIf("enabled")
                                .label("Group filter")
                                .help("Name of the OU field used to filter groups a user belongs to (optional)")
                                .optional()
                                .value(groupFilter)
                )
                ;
    }

    public static final LDAPSettings NONE = new LDAPSettings(
            false,
            "",
            "",
            "",
            "",
            "",
            "",
            "",
            "",
            ""
    );

}

File: ontrack-extension-ldap/src/main/java/net/nemerosa/ontrack/extension/ldap/LDAPSettings.kt
package net.nemerosa.ontrack.extension.ldap

import com.fasterxml.jackson.annotation.JsonProperty
import net.nemerosa.ontrack.model.annotations.APIDescription

@APIDescription("LDAP settings")
data class LDAPSettings(
        @get:JsonProperty("enabled")
        @APIDescription("Is LDAP authentication enabled?")
        val isEnabled: Boolean,
        val url: String = "",
        val searchBase: String = "",
        val searchFilter: String = "",
        val user: String = "",
        val password: String = "",
        val fullNameAttribute: String? = "",
        val emailAttribute: String? = "",
        val groupAttribute: String? = "",
        val groupFilter: String? = "",
        /**
         * The ID of the attribute which contains the name for a group
         */
        val groupNameAttribute: String? = DEFAULT_GROUP_NAME_ATTRIBUTE,
        /**
         * The base DN from which the search for group membership should be performed
         */
        val groupSearchBase: String? = DEFAULT_GROUP_SEARCH_BASE,
        /**
         * The pattern to be used for the user search. {0} is the user's DN
         */
        val groupSearchFilter: String? = DEFAULT_GROUP_SEARCH_FILTER
) {

    fun withPassword(password: String): LDAPSettings {
        return if (this.password == password) this else LDAPSettings(isEnabled, url, searchBase, searchFilter, user, password, fullNameAttribute, emailAttribute, groupAttribute, groupFilter, groupNameAttribute, groupSearchBase, groupSearchFilter)
    }

    fun withFullNameAttribute(fullNameAttribute: String): LDAPSettings {
        return if (this.fullNameAttribute == fullNameAttribute) this else LDAPSettings(isEnabled, url, searchBase, searchFilter, user, password, fullNameAttribute, emailAttribute, groupAttribute, groupFilter, groupNameAttribute, groupSearchBase, groupSearchFilter)
    }

    fun withEmailAttribute(emailAttribute: String): LDAPSettings {
        return if (this.emailAttribute == emailAttribute) this else LDAPSettings(isEnabled, url, searchBase, searchFilter, user, password, fullNameAttribute, emailAttribute, groupAttribute, groupFilter, groupNameAttribute, groupSearchBase, groupSearchFilter)
    }

    fun withGroupAttribute(groupAttribute: String): LDAPSettings {
        return if (this.groupAttribute == groupAttribute) this else LDAPSettings(isEnabled, url, searchBase, searchFilter, user, password, fullNameAttribute, emailAttribute, groupAttribute, groupFilter, groupNameAttribute, groupSearchBase, groupSearchFilter)
    }

    fun withGroupFilter(groupFilter: String): LDAPSettings {
        return if (this.groupFilter == groupFilter) this else LDAPSettings(isEnabled, url, searchBase, searchFilter, user, password, fullNameAttribute, emailAttribute, groupAttribute, groupFilter, groupNameAttribute, groupSearchBase, groupSearchFilter)
    }

    fun withGroupNameAttribute(groupNameAttribute: String): LDAPSettings {
        return if (this.groupNameAttribute == groupNameAttribute) this else LDAPSettings(isEnabled, url, searchBase, searchFilter, user, password, fullNameAttribute, emailAttribute, groupAttribute, groupFilter, groupNameAttribute, groupSearchBase, groupSearchFilter)
    }

    fun withGroupSearchBase(groupSearchBase: String): LDAPSettings {
        return if (this.groupSearchBase == groupSearchBase) this else LDAPSettings(isEnabled, url, searchBase, searchFilter, user, password, fullNameAttribute, emailAttribute, groupAttribute, groupFilter, groupNameAttribute, groupSearchBase, groupSearchFilter)
    }

    fun withGroupSearchFilter(groupSearchFilter: String): LDAPSettings {
        return if (this.groupSearchFilter == groupSearchFilter) this else LDAPSettings(isEnabled, url, searchBase, searchFilter, user, password, fullNameAttribute, emailAttribute, groupAttribute, groupFilter, groupNameAttribute, groupSearchBase, groupSearchFilter)
    }

    companion object {
        val NONE = LDAPSettings(
                isEnabled = false,
                url = "",
                searchBase = "",
                searchFilter = "",
                user = "",
                password = "",
                fullNameAttribute = "",
                emailAttribute = "",
                groupAttribute = "",
                groupFilter = "",
                groupNameAttribute = "",
                groupSearchBase = "",
                groupSearchFilter = ""
        )

        const val DEFAULT_GROUP_NAME_ATTRIBUTE = "cn"
        const val DEFAULT_GROUP_SEARCH_BASE = ""
        const val DEFAULT_GROUP_SEARCH_FILTER = "(member={0})"
    }

}